Improve OSC 8 Hyperlink parsing logic

skyline75489 opened the pull request:
This PR improves the OSC 8 Hyperlink parsing logic, by adding support to
`:` in params.

## Validation Steps Performed

Tests added & passed.

skyline75489 commented:
Pro tips: you may want to use "Split" mode to view the diff.

zadjii-msft reviewed (approved):
protip: also use the "Hide whitespace changes" option, makes this even easier to review.

DHowett commented on src/terminal/parser/OutputStateMachineEngine.cpp:
```diff
@@ -905,23 +905,24 @@ bool OutputStateMachineEngine::_ParseHyperlink(const std::wstring_view string,
-    const size_t midPos = string.find(';');
-    if (midPos != std::wstring::npos)
+    if (len != 1)
     {
-        if (len != 1)
+        uri = til::at(parts, 1);
+        const auto paramStr = til::at(parts, 0);
```
i'd like to understand what the len 1 check is doing now that we're using SplitString

skyline75489 replied:
So the basic idea I think it's for a string `;`, there will be 2 parts which are all empty strings. I think it can be a short circuit logic to return early in this case. But you're right. It can be deleted to make this even cleaner

skyline75489 replied:
@PankajBhojwani  Will it damage something if I remove the `len != 1` check? The main reason I leave it here is because I try to minimize my diffs.

PankajBhojwani replied:
With the `string == L";"` check above we don't need the `len` check anymore!

DHowett commented on src/terminal/parser/OutputStateMachineEngine.cpp:
```diff
@@ -1005,22 +1006,22 @@ bool OutputStateMachineEngine::_GetOscSetClipboard(const std::wstring_view strin
+    const std::wstring_view substr = til::at(parts, 1);
+    if (substr == L"?")
+    {
+        queryClipboard = true;
```
please drop the `else` case and un-indent the base64 return. It was hard to tell that you were deleting a redundant return.

DHowett commented on src/types/utils.cpp:
```diff
@@ -421,6 +422,11 @@ std::vector<std::wstring_view> Utils::SplitString(const std::wstring_view wstr,
+catch (...)
+{
+    LOG_CAUGHT_EXCEPTION();
+    return std::vector<std::wstring_view>();
```
```suggestion
    return {};
```

DHowett reviewed (changes requested):
-

j4james commented on src/terminal/parser/OutputStateMachineEngine.cpp:
```diff
@@ -905,23 +905,24 @@ bool OutputStateMachineEngine::_ParseHyperlink(const std::wstring_view string,
 {
     params.clear();
     uri.clear();
+    const auto parts = Utils::SplitString(string, L';');
+    if (parts.size() < 2)
```
Would this not cause problems for URLs containing a semicolon? I'm assuming you'd get more than 2 parts, and you'd then just be dropping part of the URL. Or have I misunderstood the way this works?

skyline75489 replied:
Interesting. Obviously no test case is covering the URLs containing a `;`. And yeah I think it's flawed. I'll add more cases.

skyline75489 commented:
After some consideration, I've repurposed this PR to improve the OSC 8 hyperlink parsing, instead of just trying to make use of `SplitString`. I've added the support for `:` inside the OSC 8 params and added some test cases.

PankajBhojwani reviewed (approved):
Looks good to me, thank you for this!

skyline75489 commented on src/types/ut_types/UtilsTests.cpp:
```diff
@@ -104,6 +104,8 @@ void UtilsTests::TestSplitString()
     result = SplitString(L"1", L';');
     VERIFY_ARE_EQUAL(1u, result.size());
+    result = SplitString(L";", L';');
+    VERIFY_ARE_EQUAL(2u, result.size());
```
OK the result may seem weird in a general purpose function. But it works pretty well as for parameter parsing for VT sequences.

skyline75489 commented:
A ping to stop the bot from marking the PR as stale.

DHowett commented on src/terminal/parser/OutputStateMachineEngine.cpp:
```diff
@@ -905,18 +905,24 @@ bool OutputStateMachineEngine::_ParseHyperlink(const std::wstring_view string,
+        uri = string.substr(midPos + 1);
+        const auto paramStr = string.substr(0, midPos);
+        const auto paramParts = Utils::SplitString(paramStr, ':');
+        for (const auto part : paramParts)
```
```suggestion
        for (const auto& part : paramParts)
```

DHowett reviewed (commented):
Minor things

skyline75489 commented:
Well, what do you know, the "minor" thing is causing ` error C26445: A reference to gsl::span or std::string_view may be an indication of a lifetime issue (gsl.view).`...

DHowett commented:
Hmm. That warning seems to be overzealous. We should suppress it.

skyline75489 commented:
Hey, what do you know, the tests passed like it should. @DHowett  Can I get a green light on this one?

DHowett reviewed (approved):
Thanks! Sorry -- I was worried about making some string parsing changes as it were. I kicked back for far too long on this PR.
